renku/ui/api/models/plan.py

```diff
@@ -0,0 +1,213 @@
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+r"""Renku API Plan.
+
+``Plan`` and ``CompositePlan`` classes represent Renku workflow plans executed
+in a Project. You can get a list of all active plans in a project by calling
+the ``list`` static method of each of these classes:
+
+.. code-block:: python
+
+    from renku.ui.api import Plan
+
+    plans = Plan.list()
+
+"""
+
+from datetime import datetime
+from typing import List, Optional, Union
+
+from renku.command.command_builder.database_dispatcher import DatabaseDispatcher
+from renku.core import errors
+from renku.domain_model.workflow import composite_plan as core_composite_plan
+from renku.domain_model.workflow import plan as core_plan
+from renku.infrastructure.gateway.plan_gateway import PlanGateway
+from renku.ui.api.models.project import ensure_project_context
+from renku.ui.api.models.run import Input, Link, Mapping, Output, Parameter
+
+
+class Plan:
+    """API Plan."""
+
+    def __init__(
+        self,
+        command: str,
+        date_created: Optional[datetime] = None,
+        description: Optional[str] = None,
+        inputs: List[Input] = None,
+        invalidated_at: Optional[datetime] = None,
+        keywords: Optional[List[str]] = None,
+        name: Optional[str] = None,
+        outputs: List[Output] = None,
+        parameters: List[Parameter] = None,
+        success_codes: Optional[List[int]] = None,
+    ):
+        self.command: str = command
+        self.date_created: Optional[datetime] = date_created
+        self.description: Optional[str] = description
+        self.inputs: List[Input] = inputs or []
+        self.invalidated_at: Optional[datetime] = invalidated_at
+        self.keywords: List[str] = keywords or []
+        self.name: Optional[str] = name
+        self.outputs: List[Output] = outputs or []
+        self.parameters: List[Parameter] = parameters or []
+        self.success_codes: List[int] = success_codes or []
+
+    @classmethod
+    def from_plan(cls, plan: core_plan.Plan) -> "Plan":
+        """Create an instance from a core Plan model.
+
+        Args:
+            plan(core_plan.Plan): The core plan.
+
+        Returns:
+            Plan: An API Plan model.
+        """
+        return cls(
+            command=plan.command,
+            date_created=plan.date_created,
+            description=plan.description,
+            inputs=[Input.from_parameter(i) for i in plan.inputs],
+            invalidated_at=plan.invalidated_at,
```

Comment:
Should we just keep a boolean field `deleted` in the api? Not sure `invalidated_at` is meaningful for a user. Or change to `date_deleted`.

Reply:
I've removed it and added a `deleted` field instead. We should have named it `date_deleted` in the core `Plan` class as well.